packages/SwingSet/src/vats/comms/delivery.js

```diff
@@ -236,16 +255,65 @@ export function makeDeliveryKit(state, syscall, transmit, clistKit, stateKit) {
     return { send: target, kernel: true };
   }
 
+  function resolutionCollector() {
+    const resolutions = [];
+    const doneResolutions = new Set();
+
+    function scanSlots(slots) {
+      for (const slot of slots) {
+        const { type } = parseLocalSlot(slot);
+        if (type === 'promise') {
+          const p = state.promiseTable.get(slot);
+          assert(p, X`should have a value for ${slot} but didn't`);
+          if (p.resolved && !doneResolutions.has(slot)) {
+            collect(slot);
+          }
+        }
+      }
+    }
+
+    function collect(lpid) {
+      doneResolutions.add(lpid);
+      const p = state.promiseTable.get(lpid);
+      resolutions.push([lpid, p.rejected, p.data]);
+      scanSlots(p.data.slots);
+    }
+
+    function forPromise(lpid) {
+      collect(lpid);
+      return resolutions;
+    }
+
+    function forSlots(slots) {
+      scanSlots(slots);
+      return resolutions;
+    }
+
+    return {
+      forPromise,
+      forSlots,
+    };
+  }
+
   function handleSend(localDelivery) {
     // { target, method, result, args }
     // where does it go?
     const where = resolveTarget(localDelivery.target, localDelivery.method);
 
     if (where.send) {
+      const auxResolutions = resolutionCollector().forSlots(
+        localDelivery.args.slots,
+      );
       if (where.kernel) {
         sendToKernel(where.send, localDelivery);
+        if (auxResolutions.length > 0) {
+          resolveToKernel(auxResolutions);
+        }
       } else {
         sendToRemote(where.send, where.remoteID, localDelivery);
+        if (auxResolutions.length > 0) {
+          resolveToRemote(where.remoteID, auxResolutions);
```

Comment:
I think this is correct, but it interacts with how we're going to handle the acked-retirement over comms links.

Suppose we're sending a message to remoteA, and one of the arguments references a recently-resolved promise that remoteA also knows about:

```js
alice~.one(pk1.promise);
pk1.resolve();
alice~.two(pk1.promise);
```

That should turn into three messages to remoteA:

* send(alice, "one", [rp1])
* resolve(rp1, something)  (yeah it's really a one-element batch)
* send(alice, "two", [???])

When we (comms) get `handleResolve()`, we'll mark lpid1 in our promise table as resolved, then we'll notice that remoteA is a subscriber, so we'll `resolveToRemote(remoteA, lpid1)`. What should that to do the remoteA clist? 

For the imminent retirement handshake work, I think it should remove the `lpid1 -> rpid1` mapping (the outbound side), but it should *retain* the `rpid1 -> lpid1` mapping (the inbound side). We're on the hook to accept messages from remoteA  that reference `rpid1` until we've seen the ack (so we must retain the inbound side), but we should not *send* anything that references `rpid1` (because our `resolve` message is also a retirement). So deleting the outbound side means we *can't* mention "`rpid1`", we don't even have it as a value in any table.

So then when we get to sending `two`, what rpid should it use? It needs to use a new one. The `localDelivery` we give to `handleSend` sees `lpid1`, the `resolutionCollector` includes `lpid1` in the `auxResolutions` (because it's marked as resolved in our promise table), `sendToRemote` translates the `localDelivery`'s arguments from `lpid1` into a brand-new ephemeral `rpid2`. At this point the remoteA clist looks like:

* `lpid1 -> rpid2`
* `rpid2 -> lpid1`
* `rpid1 -> lpid1` (this is waiting for the ack from remoteA)

The `send` message cites `rpid2`. Then, right after `sendToRemote` returns, `handleSend` runs `resolveToRemote(.. auxResolutions)`, which sends a `resolve` bundle that cites `rpid2`. The act of sending that bundle removes the `lpid1 -> rpid2` mapping from the clist, leaving just the inbound side, waiting for it's own ack.

Eventually, when the remote side sends those acks, we'll delete the `rpid1 -> lpid1` and `rpid2 -> lpid1` mappings, which will dereference the promise table entries for `lpid1` and `lpid2`, and might allow us to delete them (if nothing else is referencing them).

So this is good, we just need to remember the split clist thing during the next phase.

Reply:
Yes, that analysis sounds correct to me.  The auxResolves thing should do the right thing because it only cares about the state of the promise locally.  At the point we send `two`, the resolve will already be on the wire ahead of us, so can assume that by the time `two` gets there the receiving remote has forgotten the prior rpid and thus we know we need to re-export it with a new ID (and then send the immediately subsequent resolve); the asynchrony is only an issue for messages coming towards us from the other end.